activation/nipost.go

```diff
@@ -212,21 +187,24 @@ func (nb *NIPostBuilder) BuildNIPost(ctx context.Context, challenge *types.NIPos
 
 		submitCtx, cancel := context.WithDeadline(ctx, poetRoundStart)
 		defer cancel()
-		poetRequests, err := nb.submitPoetChallenges(submitCtx, poetProofDeadline, challengeHash.Bytes())
-		if err != nil {
+		if err := nb.submitPoetChallenges(submitCtx, poetProofDeadline, challenge.Hash().Bytes()); err != nil {
 			return nil, fmt.Errorf("submitting to poets: %w", err)
 		}
-		if len(poetRequests) == 0 {
+		count, err := nipost.PoetRegistrationCount(nb.localDB, nb.signer.NodeID())
+		if err != nil {
+			nb.log.Panic("failed to get poet registration count", zap.Error(err))
+		}
```

Comment:
An error here means that there is some issue that the program cannot easily recover from (SQL query cannot be parsed, DB connection dropped, etc.) and it makes sense to exit the program. I'm not a big fan of moving the panic into the function. `PoetRegistrationCount` is only called in two places and to safe this check would require passing the logger to the DB (which we do not do anywhere else).

Reply:
Right, hence the panic - I'm totally fine with it. I'm only proposing to move the panic inside the function. You don't need to pass the logger to do that - just call `panic(...)`. IMHO it makes more sense to handle bugs like a broken SQL query inside of the function rather than leaking this to the caller (who has no idea why it could fail).

Reply:
Interestingly, we seem to not have used `select count(*)` before in any other query, so there is no precedence to compare against.

I'm a bit hesitant to make `PoetRegistrationCount` directly panic on an error, because most if not all DB queries currently handle errors from `db.Exec` by just wrapping and returning them to the caller. The only way `db.Exec` however can fail is by either a) the connection to the DB was closed or b) the query is syntactically wrong. In both cases it makes no sense to continue and the node should just panic in my opinion; not just here.

I think the better way is to change `db.Exec` to not return an error and panic instead - to be consistent across all DB queries and if its OK for you I would do that in a dedicated PR.

Reply:
I created the following issue to keep track of the change: https://github.com/spacemeshos/go-spacemesh/issues/5298

Reply:
it is not the exhaustive list. it can also exist with https://www.sqlite.org/rescode.html#busy depending on how it is used

Reply:
Changed to instead return the error (so it will be logged and a bit later retried).